### ISSUE TITLE
uops.info Alder Lake-P latency for SHLX

### ISSUE THREAD

**Issue by tavianator** [andreas-abel/nanoBench#33]
https://uops.info/html-instr/SHLX_R64_R64_R64.html#ADL-P lists `SHLX` as having 3-cycle latency for both operands.  This is in contrast to Intel's docs and [InstLatx64](http://users.atw.hu/instlatx64/)'s measurements.  So what gives?

I looked into it and figured out something strange.  If you run the [specified nanoBench command](https://uops.info/html-lat/ADL-P/SHLX_R64_R64_R64-Measurements.html#lat2-%3E1), you'll see

```console
# lscpu
...
  Model name:             12th Gen Intel(R) Core(TM) i7-1280P
...
# ./nanoBench.sh -f -unroll 100 -warm_up_count 10 -config configs/cfg_AlderLakeP_common.txt -cpu 0 \
    -asm "SHLX R9, R8, R10; MOVSX R8, R9D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D" \
...
Core cycles: 8.00
...
```

Here R10 has some arbitrary value.  Let's try setting it to 1:

```console
# ./nanoBench.sh -f -unroll 100 -warm_up_count 10 -config configs/cfg_AlderLakeP_common.txt -cpu 0 \
    -asm "SHLX R9, R8, R10; MOVSX R8, R9D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D" \
    -asm_init "MOV R10, 1"
...
Core cycles: 8.00
...
# ./nanoBench.sh -f -unroll 100 -warm_up_count 10 -config configs/cfg_AlderLakeP_common.txt -cpu 0 \
    -asm "SHLX R9, R8, R10; MOVSX R8, R9D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D; MOVSX R8, R8D" \
    -asm_init "MOV R10D, 1"
...
Core cycles: 6.00
...
```

So it actually matters *how* the count register is initialized.  It seems like if it's a 64-bit op, then you get 3c latency.  But a 32-bit op that implicitly zeroes the top half gets it down to 1c latency.  Other things like `MOV R10D, R10D` also work.

I'm not sure how the nanoBench commands on uops.info are generated, so I'm reporting this here.

Relevant discussion here:

- https://x.com/HaroldAptroot/status/1873461353203302819
- https://bsky.app/profile/haroldaptroot.bsky.social/post/3leht4lp2dk2u
- https://stackoverflow.com/a/77705726/502399

**Comment by tavianator** [comment]
More details here: https://tavianator.com/2025/shlxplained.html

It's related to the "small immediate add renaming" optimization introduced in Alder Lake

**Comment by ValeZAA** [comment]
> More details here: https://tavianator.com/2025/shlxplained.html
> 
> It's related to the "small immediate add renaming" optimization introduced in Alder Lake

Are you sure that latency is 0.20? Not zero until reorder buffer ends? What happens if you execute 4 instructions, does it have latency 0.20 * 4 or zero after all? What happens if you do 6 instructions, does it have 0.20 * 5 + 0.20 or 2 cycles?

**Comment by tavianator** [comment]
> > More details here: https://tavianator.com/2025/shlxplained.html
> > It's related to the "small immediate add renaming" optimization introduced in Alder Lake
> 
> Are you sure that latency is 0.20? Not zero until reorder buffer ends? What happens if you execute 4 instructions, does it have latency 0.20 * 4 or zero after all? What happens if you do 6 instructions, does it have 0.20 * 5 + 0.20 or 2 cycles?

Fair enough, what I meant by that was that it can execute 5 dependent `add rax, 1` instructions per cycle.  Interestingly it can also only do 5 `inc rax` per cycle, even though it should be able to decode 6 per cycle (right?)

Within the frontend limits, it's 0 latency.  It can do `6*inc rcx + 3*movsx rcx, ecx` in 3c, but `7*inc rcx` pushes it to 3.04c.

**Comment by ValeZAA** [comment]
>that it can execute 5 dependent add rax, 1 instructions per cycle. 

That means 0.20 latency. Last time something like this was possible was Pentium 4 Netburst, specifically Netburst Northwood. https://stackoverflow.com/questions/45066299/was-there-a-p4-model-with-double-pumped-64-bit-operations

>Interestingly it can also only do 5 inc rax per cycle, even though it should be able to decode 6 per cycle (right?)

12 ports and 6 wide decode do not matter, because only 5 integer ALU ports exist, 1 was added, still 6 wide decode means there is 1 uop that you can insert until the end of cycle. https://stackoverflow.com/questions/21819682/is-integer-multiplication-really-done-at-the-same-speed-as-addition-on-a-modern#comment131570325_57441231

Peter Cordes argued to me is zero cycles spend on this 11 bit counter:

"My point is it's not some non-zero fractional number, it's actually 0 cycles. And you can interleave inc / add imm8 into a dependency chain of known length to confirm that, like nanobench does on uops.info to confirm that. The article you linked mentions the anomalously high latency measured for shrx using such a microbenchmark." That is your article.

**Comment by tavianator** [comment]
> 12 ports and 6 wide decode do not matter, because only 5 integer ALU ports exist

I don't think the execution ports matter, if you check the counters you'll see no uops are being retired at all for those instructions (just like an eliminated mov).

And everyone agrees that an eliminated mov has 0 latency right?  This is not really any different.  Even though the CPU cannot execute $\infty$ of them per second, due to frontend
limits.

**Comment by andreas-abel** [comment]
I have added experiments that initialize the upper half of the registers with 0. 

Other instructions that were affected by this: BEXTR, BZHI, SARX, and VPBROADCAST.

**Comment by tavianator** [comment]
Huh I'm surprised by VPBROADCAST
